src/authorization/WebAclReader.ts

```diff
@@ -147,122 +129,115 @@ export class WebAclReader extends PermissionReader {
+      for (const target of targets) {
+        if (!longest.path.includes(target.path)) {
+          continue;
+        }
+
+        // Store all matches in the matchingIdentifiers array and remove them from the list of targets.
+        // Only need to check length since we already know both paths are a subpath of `long`.
+        if (target.path.length >= aclMatch.identifier.path.length) {
+          this.logger.debug(`Found ${target.path} ACL information in the ACL resource of ${aclMatch.identifier.path}`);
+          matchingIdentifiers.push(target);
+          targets.delete(target);
+        }
+      }
     }
-    return {
-      ...aclPermissions,
-      create: aclPermissions.write,
-      delete: aclPermissions.write,
-    };
+
+    return storeMap;
   }
 
   /**
-   * Finds the ACL data relevant for its resource, and potentially its parent if required.
-   * All quads in the resulting store(s) can be interpreted as being relevant ACL rules for their target.
+   * For every store/identifier combination it finds the relevant ACL triples for that identifier.
+   * This is done in such a way that store results are reused for all matching identifiers.
+   * The split is based on the `acl:accessTo` and `acl:default` triples.
    *
-   * @param target - Target to find ACL data for.
-   * @param includeParent - If parent ACL data is also needed.
-   *
-   * @returns The relevant triples.
+   * @param map - Map of matches that need to be filtered.
    */
-  private async getAcl(target: ResourceIdentifier, includeParent: boolean): Promise<AclSet> {
-    this.logger.debug(`Searching ACL data for ${target.path}${includeParent ? 'and its parent' : ''}`);
-    const to = includeParent ? this.identifierStrategy.getParentContainer(target) : target;
-    const acl = await this.getAclRecursive(target, to);
-
-    // The only possible case where `acl` has 2 values instead of 1
-    // is when the `target` has an acl, and `includeParent` is true.
-    const keys = Object.keys(acl);
-    if (keys.length === 2) {
-      const result: AclSet = { targetAcl: await this.filterStore(acl[target.path], target.path, true) };
-      // The other key will be the parent
-      const parentKey = keys.find((key): boolean => key !== target.path)!;
-      result.parentAcl = await this.filterStore(acl[parentKey], parentKey, parentKey === to.path);
-
-      return result;
-    }
-
-    // Only 1 key: no parent was requested, target had no direct acl resource, or both
-    const [ path, store ] = Object.entries(acl)[0];
-    const result: AclSet = { targetAcl: await this.filterStore(store, path, path === target.path) };
-    if (includeParent) {
-      // In case the path is not the parent, it will also just use the defaults just like the target
-      result.parentAcl = path === to.path ? await this.filterStore(store, path, true) : result.targetAcl;
+  private async filterAclMap(map: Map<AclMatch, ResourceIdentifier[]>): Promise<Map<Store, ResourceIdentifier[]>> {
```

Comment:
```suggestion
  private async findAuthorizationStatements(map: Map<AclMatch, ResourceIdentifier[]>): Promise<Map<Store, ResourceIdentifier[]>> {
```
(and, as per my remark above, maybe this is the place where the ACL document should actually be read)

Reply:
(ah and then we can even do streaming filtering, how cool is that)